### ISSUE TITLE
Should global validators run AFTER field validators?

### ISSUE THREAD

**Issue by VictorioBerra** [DataTables/Editor-NET#26]
Maybe im approaching this wrong but I have a global validator that needs to do some special custom checks.

Because my global validator runs first im duplicating a ton of field validators:

```csharp
// The entire purpose of this validator is to check if the FieldDefinition already exists associated to the same entities
// For example a field definition for Name cannot exist twice both added to InventoryServers
.Validator((editor, action, data) =>
{
    if (action == DtRequest.RequestTypes.DataTablesGet)
    {
        return string.Empty;
    }

    foreach (var pair in data.Data)
    {
        var values = pair.Value as Dictionary<string, object>;

        if (!values.TryGetValue("FieldDefinition", out var fieldDefinitionData))
        {
            // DUPLICATED FIELD VALIDATOR
            return string.Empty;
        }
        else
        {
            if (action is DtRequest.RequestTypes.EditorEdit
                or DtRequest.RequestTypes.EditorCreate)
            {
                var fieldDefinitionDictionary = fieldDefinitionData as Dictionary<string, object>;
                if (!fieldDefinitionDictionary.TryGetValue("Name", out var fieldDefinitionNameData))
                {
                    // DUPLICATED FIELD VALIDATOR
                    // An edit does not have to result in a name change (they didnt send Name field).
                    // A create Name is required, so another validator will catch this
                    return string.Empty;
                }

                var fieldDefinitionName = fieldDefinitionNameData.ToString();
                if (string.IsNullOrWhiteSpace(fieldDefinitionName))
                {
                    // DUPLICATED FIELD VALIDATOR
                    // Field validator will catch it
                    return string.Empty;
                }

                if (!values.TryGetValue("FieldDefinitionEntity", out var entities))
                {
                    // DUPLICATED FIELD VALIDATOR
                    return string.Empty;
                }

                if (action is DtRequest.RequestTypes.EditorEdit)
                {
                    var existingFieldDefinition = context.WuitFieldDefinition
                        .Include(x => x.Entities)
                        .FirstOrDefault(x => x.Id == Convert.ToInt32(pair.Key, CultureInfo.InvariantCulture));
                    if (fieldDefinitionName == existingFieldDefinition.Name)
                    {
                        // Name did not change, no need to check if it exists
                        continue;
                    }
                }

                var entitiesDict = entities as Dictionary<string, object>;
                var entityIds = entitiesDict.Values
                    .Select(x => x as Dictionary<string, object>)
                    .Select(x => (int)x["Id"])
                    .ToArray();

                if (entityIds.Count == 0)
                {
                    // DUPLICATED FIELD VALIDATOR
                    // Field validator for the MJoin below will check that they provided a min of 1
                    return string.Empty;
                }

                var existingRecord = IsFieldDefinitionInUse(fieldDefinitionName, entityIds);

                if (existingRecord is null)
                {
                    return string.Empty;
                }

                return $"Field Definition already exists for field name {fieldDefinitionName} ({existingRecord.Id}) on entities {string.Join(", ", existingRecord.Entities.Select(x => x.Name))}";
            }
        }
    }

    return string.Empty;
})
```

Now this would be my code if I could trust that field validators all ran and that it was safe to access my dictionary:

```csharp
.Validator((editor, action, data) =>
{
    if (action == DtRequest.RequestTypes.DataTablesGet)
    {
        return string.Empty;
    }

    // Create/Edit can accept multiple records at once
    foreach (var pair in data.Data)
    {
        var values = pair.Value as Dictionary<string, object>;

        if (!values.TryGetValue("FieldDefinition", out var fieldDefinitionData))
        {
            return string.Empty;
        }
        else
        {
            if (action is DtRequest.RequestTypes.EditorEdit or DtRequest.RequestTypes.EditorCreate)
            {
                var fieldDefinitionDictionary = fieldDefinitionData as Dictionary<string, object>;
                var fieldDefinitionName = fieldDefinitionDictionary["Name"].ToString();
                var entities = values["FieldDefinitionEntity"] as Dictionary<string, object>;
                var entityIds = entities.Values
                    .Select(x => x as Dictionary<string, object>)
                    .Select(x => (int)x["Id"])
                    .ToArray();

                var existingFieldDefinition = context.WuitFieldDefinition
                    .Include(x => x.Entities)
                    .FirstOrDefault(x => x.Id == Convert.ToInt32(pair.Key, CultureInfo.InvariantCulture) &&
                    x.Name == fieldDefinitionName);

                if (existingFieldDefinition is not null)
                {
                    // Name did not change, no need for further checks
                    continue;
                }

                var existingRecord = IsFieldDefinitionInUse(fieldDefinitionName, entityIds);
                if (existingRecord is null)
                {
                    return string.Empty;
                }

                return $"Field Definition already exists for field name {fieldDefinitionName} ({existingRecord.Id}) on entities {string.Join(", ", existingRecord.Entities.Select(x => x.Name))}";
            }
        }
    }

    return string.Empty;
})
```

Looking at the code looks like PreX events run before field validation too... so those are not a workaround for me either.

**Comment by VictorioBerra** [comment]
Thinking on this more, I think this amounts to a feature request:

`.PreValidate()` and `.PostValidate()` you could even leave `.PreValidate()` as `.Validate()` to be backwards compatible.

Thoughts?

**Comment by AllanJard** [comment]
Hi - thanks for raising this and illustrating it with an example. I can absolutely see how it might be useful to have the field validation run first, and then stop if there is an error. If I recall correctly, my intention with running the global validation first had been to allow a quick check for access rights.

Perhaps an overload to `Editor.Validate()` to tell it to run the validator after the fields (or not - default) might be the way to do this, or an `Editor.ValidateAfterFields()` (ugly!) method.

I'll set a 2.4 target for this as I think it would be a good addition.

**Comment by VictorioBerra** [comment]
An overload would be awesome! Thanks Allan

**Comment by AllanJard** [comment]
I've [added this ability in here](https://github.com/DataTables/Editor-NET/commit/9797df1f81d7d383c2e9d110e51c1a59045f1b50).

This is the overload signature:

```cs
editor.Validator( );          // get before field validation validators
editor.Validator( fn );       // validator for before field validation
editor.Validator( true, fn ); // validator for after field validation
editor.Validator( true );     // get after field validation validators
```

`false` can be used as the first argument for final two overloads, and it will act as the first two would.

This will be included in the up coming 2.4 release :). Thanks for the suggestion!

**Comment by VictorioBerra** [comment]
Thank you!
